### PR TITLE
Adding typechecks to Column Operations

### LINKED ISSUE #6106: Add missing type-checks to column operations
With #5158 we have a reliable source of type information for columns. Some operations are already taking advantage of this and checking if the operation makes sense for a given input column, but many operations still do not.

- [x] Review operations
	- [x] Ensure that types are checked (so that situations like `starts_with` on `Integer` column fail early with `Invalid_Value_Type` (or some variation of this error).
	- [ ] ~~Verify the logic for inferring return types in the in-memory backend.~~
		- Database already infers the types from the Database engine.
		- Ensure stuff like adding two fixed-length texts results in a fixed-length text whose length is the sum of the lengths of the terms.
			- This can be reliably tested only after #5159 is implemented.
	    - This is done partially, will be finished as part of #5159.
- [x] Add tests for these scenarios.

### PR DESCRIPTION
**Pull request by radeusgd**
### Pull Request Description

Closes #6106 

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [ ] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [ ] If GUI codebase was changed, the GUI was tested when built using `./run ide build`.

### CODE REVIEW

File: distribution/lib/Standard/Table/0.0.0-dev/src/Data/Type/Value_Type.enso [hunk 1]
```diff
@@ -186,23 +186,93 @@ type Value_Type
...
+        if bucket_1.is_nothing || bucket_2.is_nothing then False else
+            if (type_1 == Value_Type.Mixed) || (type_2 == Value_Type.Mixed) then True else
+                bucket_1 == bucket_2
+
+
+    ## PRIVATE
+       Checks that both types can be compared for ordering and runs the
+       following action or reports an error.
+
+       The second argument may be `Nothing` if the type represents null values.
```
Comment (GregoryTravis) [inline distribution/lib/Standard/Table/0.0.0-dev/src/Data/Type/Value_Type.enso]: Can you clarify what you mean by "represents null values"? Does this describe certain types but not others? Or is it about the possibility that `type_1` is Nothing?

File: distribution/lib/Standard/Table/0.0.0-dev/src/Data/Type/Value_Type_Helpers.enso [hunk 2]
```diff
@@ -75,3 +78,95 @@ find_common_type types strict =
...
+            Enso_Types.most_specific_value_type value use_smallest=True
+
+    ## PRIVATE
+       A helper that can be used to check the argument value type, while allowing
+       `Nothing` values too.
+
+       > Example
+
+         type_helpers.check_argument_type other Value_Type.expect_text <| operation
+    check_argument_type self value type_check ~action =
```
Comment (GregoryTravis) [inline distribution/lib/Standard/Table/0.0.0-dev/src/Data/Type/Value_Type_Helpers.enso]: Could you add a type signature here?

Comment (GregoryTravis) [reply distribution/lib/Standard/Table/0.0.0-dev/src/Data/Type/Value_Type_Helpers.enso]: Also it might be good to mention that `type_check` is an assertion that can throw an error, rather than a predicate.

**Review by GregoryTravis — APPROVED** [review]

File: distribution/lib/Standard/Table/0.0.0-dev/src/Data/Type/Value_Type.enso (hunk 1 above)
Comment (radeusgd) [reply distribution/lib/Standard/Table/0.0.0-dev/src/Data/Type/Value_Type.enso]: It's the latter, I will try to rephrase this to be clearer.

File: distribution/lib/Standard/Database/0.0.0-dev/src/Internal/Database_Type_Helpers.enso [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+
+from Standard.Table.Data.Type.Value_Type_Helpers import Operation_Type_Helpers
+
+import project.Data.Column.Column
+
+## PRIVATE
+   Type helpers for database tables.
+type_helpers =
+    is_column value = value.is_a Column
+    Operation_Type_Helpers.Instance is_column
```
Comment (jdunkerley) [inline distribution/lib/Standard/Database/0.0.0-dev/src/Internal/Database_Type_Helpers.enso]: No idea what this is for could we have a little more detail on the doc string please.

File: distribution/lib/Standard/Table/0.0.0-dev/src/Errors.enso [hunk 4]
```diff
@@ -417,7 +417,7 @@ type Invalid_Value_Type
     to_display_text self =
         prefix = case self.related_column of
             Nothing -> "Expected "
-            column_name -> "Expected " + column_name + " column to have "
+            column_name : Text -> "Expected " + column_name + " column to have "
```
Comment (jdunkerley) [inline distribution/lib/Standard/Table/0.0.0-dev/src/Errors.enso]: Can we add a catch all here please. I know we shouldn't get here but throwing an error in the `to_display_text` will mean we just can't see anything in IDE if something set wrongly. (e.g. we passed a Column instead of the name)

File: distribution/lib/Standard/Table/0.0.0-dev/src/Errors.enso [hunk 5]
```diff
@@ -451,16 +451,26 @@ type Column_Type_Mismatch
...
 
        Create a human-readable version of the error.
     to_display_text : Text
     to_display_text self =
-        "No common type could have been found for the columns corresponding to ["+self.column_name+"]. If you want to allow mixed types, please retype the columns to the `Mixed` before the concatenation (note however that most Database backends do not support `Mixed` types, so it may work only for the in-memory backend)."
+        types = self.types.map .to_display_text . join ", "
+        prefix = "No common type could have been found for the types: "+types
+        infix = case self.related_column_name of
+            Nothing -> "."
+            column_name : Text -> " when unifying the column ["+column_name+"]."
```
Comment (jdunkerley) [inline distribution/lib/Standard/Table/0.0.0-dev/src/Errors.enso]: as above.

File: distribution/lib/Standard/Table/0.0.0-dev/src/Errors.enso [hunk 6]
```diff
@@ -451,16 +451,26 @@ type Column_Type_Mismatch
...
-       An error indicating that no common type could be found for the merged
-       columns.
-    Error (column_name : Text)
+       An error indicating that no common type could be found.
+
+       Arguments:
+       - types: The types that were tried to be unified.
+       - related_column_name: The name of the resulting column that was being
+         unified, if applicable.
+    Error (types : Vector Value_Type) (related_column_name : Nothing|Text)
```
Comment (jdunkerley) [inline distribution/lib/Standard/Table/0.0.0-dev/src/Errors.enso:458]: should we default `related_column_name` to `Nothing`?

File: distribution/lib/Standard/Database/0.0.0-dev/src/Internal/Statement_Setter.enso [hunk 7]
```diff
@@ -35,4 +35,10 @@ fill_hole_default stmt i value = case value of
     _ : Decimal   -> stmt.setDouble i value
     _ : Text      -> stmt.setString i value
     _ : Date_Time -> JDBCUtils.setZonedDateTime stmt i value
-    _             -> stmt.setObject i value
+    ## Time_Of_Day and Date_Time sometimes work ok, but sometimes are passed as
```
Comment (jdunkerley) [inline distribution/lib/Standard/Database/0.0.0-dev/src/Internal/Statement_Setter.enso]: `Date_Time` should be `Date`?

File: distribution/lib/Standard/Table/0.0.0-dev/src/Data/Type/Value_Type.enso [hunk 8]
```diff
@@ -186,23 +186,97 @@ type Value_Type
         Value_Type.Date_Time _ -> True
         _ -> False
 
+    ## PRIVATE
+       Specifies if values of the given type can be compared for ordering.
+    has_ordering : Boolean
+    has_ordering self = case self of
+        Value_Type.Integer _ -> True
```
Comment (jdunkerley) [inline distribution/lib/Standard/Table/0.0.0-dev/src/Data/Type/Value_Type.enso]: Why not `Value_Type.Byte`?

Comment (jdunkerley) [reply distribution/lib/Standard/Table/0.0.0-dev/src/Data/Type/Value_Type.enso]: and I think `Boolean` has an order both in Enso and in DB (`False < True`)

File: test/Table_Tests/src/In_Memory/Table_Spec.enso [hunk 9]
```diff
@@ -193,7 +196,7 @@ spec =
 
             x = Column.from_vector 'a' [1, Nothing, 2, Nothing]
             y = Column.from_vector 'b' [0, 0, 0, 0]
-            r = x.zip y skip_missing=False x-> y-> if x == 1 then x else y
+            r = x.zip y skip_nothing=False x-> y-> if x == 1 then x else y
```
Comment (jdunkerley) [inline test/Table_Tests/src/In_Memory/Table_Spec.enso:199]: Good spot. Can we do a quick scan for where we still have `missing` in the code base as we should align I think.

**Review by jdunkerley — COMMENTED** [review]
Through quite a lot of the files... will finish review shortly but here are a few comments!

File: distribution/lib/Standard/Table/0.0.0-dev/src/Data/Type/Value_Type.enso (hunk 8 above)
Comment (radeusgd) [reply distribution/lib/Standard/Table/0.0.0-dev/src/Data/Type/Value_Type.enso]: Byte - sure, I kind of thought of it as a binary format and not a number, but obviously that is wrong - I will amend.

As for Booleans - I remember a discussion long time ago that ordering booleans does not make sense so Enso should not support this. But I think the decision has been changed since and indeed, Booleans are comparable in Enso now. Then I can add them here as well.

File: distribution/lib/Standard/Table/0.0.0-dev/src/Errors.enso (hunk 6 above)
Comment (radeusgd) [reply distribution/lib/Standard/Table/0.0.0-dev/src/Errors.enso:458]: I thought it could be better to keep it required and specify `Nothing` explicitly. Otherwise, with other errors like that I was sometimes _not adding_ the related column, because it was optional and I was simply forgetting it is supposed to be there. Whereas, if it's a required argument it forces to think if we can 'put something' there.

File: test/Table_Tests/src/In_Memory/Table_Spec.enso (hunk 9 above)
Comment (radeusgd) [reply test/Table_Tests/src/In_Memory/Table_Spec.enso:199]: A quick grep for `_missing` yields only stuff like:
- `error_on_missing_columns` - that should probably stay
- `if_missing` in operations like `find` or `get` - that sounds good as it
- one example which should be renamed.

Comment (radeusgd) [reply test/Table_Tests/src/In_Memory/Table_Spec.enso:199]: Oh, and the `count_missing` method! I guess that one could be renamed.

File: distribution/lib/Standard/Database/0.0.0-dev/src/Data/Table.enso [hunk 10]
```diff
@@ -1523,6 +1518,53 @@ type Table
...
+         Remove leading and trailing spaces from cells in multiple columns.
+
+             table.replace_text By_Name ["foo", "bar"] "^\s*(.*?)\s*$" "$1" use_regex=True
+
+       > Example
+         Replace texts in quotes with parentheses in column at index 1.
+
+             table.replace_text 1 '"(.*?)"' '($1)' use_regex=True
+    replace_text : Text | Integer | Column_Selector | Vector (Integer | Text | Column_Selector) -> Text -> Text -> Case_Sensitivity -> Boolean -> Boolean -> Problem_Behavior -> Table
+    replace_text self columns=[0] term="" new_text="" case_sensitivity=Case_Sensitivity.Sensitive only_first=False use_regex=False on_problems=Problem_Behavior.Report_Warning =
```
Comment (jdunkerley) [inline distribution/lib/Standard/Database/0.0.0-dev/src/Data/Table.enso]: lets remove this as have removed from the in-memory please.

File: distribution/lib/Standard/Database/0.0.0-dev/src/Data/Column.enso [hunk 11]
```diff
@@ -1028,35 +1059,20 @@ type Column
     ## PRIVATE
     naming_helpers self = self.connection.dialect.get_naming_helpers
 
+
```
Comment (jdunkerley) [inline distribution/lib/Standard/Database/0.0.0-dev/src/Data/Column.enso]: why the double space?

**Review by jdunkerley — APPROVED** [review]
One slight change please.

Looks really good and great range of test coverage.

File: distribution/lib/Standard/Database/0.0.0-dev/src/Data/Column.enso (hunk 11 above)
Comment (radeusgd) [reply distribution/lib/Standard/Database/0.0.0-dev/src/Data/Column.enso]: accident

File: distribution/lib/Standard/Database/0.0.0-dev/src/Data/Table.enso (hunk 10 above)
Comment (radeusgd) [reply distribution/lib/Standard/Database/0.0.0-dev/src/Data/Table.enso]: Sure, I added it before it was removed 😅
